src/runtime-rs/crates/hypervisor/src/device/driver/virtio_fs.rs

```diff
@@ -72,7 +77,85 @@ pub struct ShareFsDeviceConfig {
     pub options: Vec<String>,
 }
 
+#[derive(Clone, Debug)]
+pub struct ShareFsConfig {
+    /// shared_path: the upperdir of the passthrough fs exported dir or rafs meta file of rafs
+    pub shared_path: String,
+
+    /// sharefs device option type: Add and Update.
+    pub sharefs_op: ShareFsDeviceOp,
+
+    /// device config for sharefs device
+    pub device_config: Option<ShareFsDeviceConfig>,
+
+    /// mount config for sharefs mount/umount/update
+    pub mount_config: Option<ShareFsMountConfig>,
+}
+
 #[derive(Debug, Clone)]
 pub struct ShareFsDevice {
-    pub config: ShareFsDeviceConfig,
+    /// device id for sharefs device in device manager
+    pub device_id: String,
+
+    /// config for sharefs device
+    pub config: ShareFsConfig,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ShareFsDeviceOp {
+    /// Add for sharefs device with add_device()
+    Add,
+
+    /// Update for sharefs mount with update_device()
+    Update,
+}
+
+impl ShareFsDevice {
+    // new creates a share-fs device
+    pub fn new(device_id: &str, config: &ShareFsConfig) -> Self {
+        Self {
+            device_id: device_id.to_string(),
+            config: config.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl Device for ShareFsDevice {
+    async fn attach(&mut self, h: &dyn hypervisor) -> Result<()> {
+        match self.config.sharefs_op {
+            ShareFsDeviceOp::Add => h
+                .add_device(DeviceType::ShareFs(self.clone()))
+                .await
+                .context("add share-fs device.")?,
+            ShareFsDeviceOp::Update => h
```

Comment:
It's kind of weird that the update operation is still handled inside `attach()`.

Reply:
fixed